src/main/java/net/fabricmc/accesswidener/AccessWidenerReader.java

```diff
@@ -57,73 +63,141 @@ public void read(BufferedReader reader, String currentNamespace) throws IOExcept
 
 		String line;
 
+		Pattern delimiter = version < V2 ? V1_DELIMITER : V2_DELIMITER;
+
 		while ((line = reader.readLine()) != null) {
 			lineNumber++;
 
-			//Comment handling
-			int commentPos = line.indexOf('#');
+			line = handleComment(version, line);
 
-			if (commentPos >= 0) {
-				line = line.substring(0, commentPos).trim();
+			if (line.isEmpty()) {
+				continue;
 			}
 
-			if (line.isEmpty()) continue;
-
 			if (Character.isWhitespace(line.codePointAt(0))) {
 				throw error("Leading whitespace is not allowed");
 			}
 
-			String[] split = line.split("\\s+");
+			// Note that this trims trailing spaces. See the docs of split for details.
+			List<String> tokens = Arrays.asList(delimiter.split(line));
+
+			boolean global = false;
 
-			AccessType access = readAccessType(split[0]);
+			if (version >= V2) {
+				// Global access widener flag
+				if (!tokens.isEmpty() && tokens.get(0).equals("global")) {
+					tokens = tokens.subList(1, tokens.size());
+					global = true;
+				}
 
-			if (split.length < 2) {
-				throw error("Expected <class|field|method> following " + split[0]);
+				// Interface injection
+				if (!tokens.isEmpty() && tokens.get(0).equals("add-interface")) {
+					handleInterfaceInjection(tokens, global);
+					continue;
+				}
```

Comment:
I would move all of this out to be done later, concenrating on one thing at once is better imo.

Reply:
Done